File: packages/core/src/marginalia_core/domain/reading_session.py
"""Reading session and state models."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime
from enum import Enum


def _utc_now() -> datetime:
    return datetime.now(UTC)


class PlaybackState(str, Enum):
    """Low-level playback status."""

    STOPPED = "stopped"
    PLAYING = "playing"
    PAUSED = "paused"


class ReaderState(str, Enum):
    """High-level lifecycle state for the local reading engine."""

    IDLE = "IDLE"
    READING = "READING"
    PAUSED = "PAUSED"
    LISTENING_FOR_COMMAND = "LISTENING_FOR_COMMAND"
    RECORDING_NOTE = "RECORDING_NOTE"
    PROCESSING_REWRITE = "PROCESSING_REWRITE"
    READING_REWRITE = "READING_REWRITE"
    ERROR = "ERROR"


@dataclass(slots=True)
class ReadingPosition:
    """Anchor inside a document."""

    section_index: int = 0
    chunk_index: int = 0
    char_offset: int = 0

    @property
    def anchor(self) -> str:
        return f"section:{self.section_index}/chunk:{self.chunk_index}"


@dataclass(slots=True)
class ReadingSession:
    """Mutable session record for CLI-driven local reading."""

    session_id: str
    document_id: str
    state: ReaderState = ReaderState.IDLE
    playback_state: PlaybackState = PlaybackState.STOPPED
    position: ReadingPosition = field(default_factory=ReadingPosition)
    active_note_id: str | None = None
    last_command: str | None = None
    last_command_source: str | None = None
    last_recognized_command: str | None = None
    voice: str | None = None
    tts_provider: str | None = None
    command_stt_provider: str | None = None
    playback_provider: str | None = None
    audio_reference: str | None = None
    playback_process_id: int | None = None
    updated_at: datetime = field(default_factory=_utc_now)

    def touch(self) -> None:
        self.updated_at = _utc_now()

File: crates/marginalia-core/src/domain/reading_session.rs
use chrono::{DateTime, Utc};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PlaybackState {
    Stopped,
    Playing,
    Paused,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ReaderState {
    Idle,
    Reading,
    Paused,
    ListeningForCommand,
    RecordingNote,
    ProcessingRewrite,
    ReadingRewrite,
    Error,
}

impl PlaybackState {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Stopped => "stopped",
            Self::Playing => "playing",
            Self::Paused => "paused",
        }
    }
}

impl ReaderState {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Idle => "idle",
            Self::Reading => "reading",
            Self::Paused => "paused",
            Self::ListeningForCommand => "listening_for_command",
            Self::RecordingNote => "recording_note",
            Self::ProcessingRewrite => "processing_rewrite",
            Self::ReadingRewrite => "reading_rewrite",
            Self::Error => "error",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct ReadingPosition {
    pub section_index: usize,
    pub chunk_index: usize,
    pub char_offset: usize,
}

impl ReadingPosition {
    pub fn anchor(&self) -> String {
        format!("section:{}/chunk:{}", self.section_index, self.chunk_index)
    }

    pub fn from_anchor(anchor: &str) -> Self {
        let mut position = Self::default();

        for item in anchor.split('/') {
            let mut parts = item.splitn(2, ':');
            let key = parts.next().unwrap_or_default();
            let value = parts.next().unwrap_or_default();

            match key {
                "section" => {
                    if let Ok(parsed) = value.parse::<usize>() {
                        position.section_index = parsed;
                    }
                }
                "chunk" => {
                    if let Ok(parsed) = value.parse::<usize>() {
                        position.chunk_index = parsed;
                    }
                }
                _ => {}
            }
        }

        position
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ReadingSession {
    pub session_id: String,
    pub document_id: String,
    pub state: ReaderState,
    pub playback_state: PlaybackState,
    pub position: ReadingPosition,
    pub active_note_id: Option<String>,
    pub last_command: Option<String>,
    pub last_command_source: Option<String>,
    pub last_recognized_command: Option<String>,
    pub voice: Option<String>,
    pub tts_provider: Option<String>,
    pub command_stt_provider: Option<String>,
    pub playback_provider: Option<String>,
    pub command_listening_active: bool,
    pub command_language: Option<String>,
    pub audio_reference: Option<String>,
    pub playback_process_id: Option<u32>,
    pub runtime_process_id: Option<u32>,
    pub runtime_status: Option<String>,
    pub runtime_error: Option<String>,
    pub startup_cleanup_summary: Option<String>,
    pub is_active: bool,
    pub updated_at: DateTime<Utc>,
}

impl ReadingSession {
    pub fn new(session_id: impl Into<String>, document_id: impl Into<String>) -> Self {
        Self {
            session_id: session_id.into(),
            document_id: document_id.into(),
            state: ReaderState::Idle,
            playback_state: PlaybackState::Stopped,
            position: ReadingPosition::default(),
            active_note_id: None,
            last_command: None,
            last_command_source: None,
            last_recognized_command: None,
            voice: None,
            tts_provider: None,
            command_stt_provider: None,
            playback_provider: None,
            command_listening_active: false,
            command_language: None,
            audio_reference: None,
            playback_process_id: None,
            runtime_process_id: None,
            runtime_status: None,
            runtime_error: None,
            startup_cleanup_summary: None,
            is_active: true,
            updated_at: Utc::now(),
        }
    }

    pub fn touch(&mut self) {
        self.updated_at = Utc::now();
    }
}

#[cfg(test)]
mod tests {
    use super::{PlaybackState, ReaderState, ReadingPosition, ReadingSession};

    #[test]
    fn reading_position_round_trips_anchor() {
        let position = ReadingPosition {
            section_index: 3,
            chunk_index: 7,
            char_offset: 2,
        };

        assert_eq!(position.anchor(), "section:3/chunk:7");
        assert_eq!(ReadingPosition::from_anchor(&position.anchor()).section_index, 3);
        assert_eq!(ReadingPosition::from_anchor(&position.anchor()).chunk_index, 7);
    }

    #[test]
    fn reading_session_new_uses_expected_defaults() {
        let session = ReadingSession::new("session-1", "doc-1");

        assert_eq!(session.session_id, "session-1");
        assert_eq!(session.document_id, "doc-1");
        assert_eq!(session.state, ReaderState::Idle);
        assert_eq!(session.playback_state, PlaybackState::Stopped);
        assert!(session.is_active);
        assert!(!session.command_listening_active);
    }
}
